rust/src/csv/reader.rs

```diff
@@ -54,6 +57,132 @@ use crate::record_batch::RecordBatch;
+    if string.starts_with("\"") {
+        return DataType::Utf8;
+    }
+    // match regex in a particular order
+    if BOOLEAN_RE.is_match(string) {
+        return DataType::Boolean;
+    } else if DECIMAL_RE.is_match(string) {
+        return DataType::Float64;
+    } else if INTEGER_RE.is_match(string) {
+        return DataType::Int64;
+    } else {
+        return DataType::Utf8;
+    }
+}
+
+/// Infer the schema of a CSV file by reading through the first n records of the file,
+/// with `max_read_records` controlling the maximum number of records to read.
+///
+/// If `max_read_records` is not set, the whole file is read to infer its schema.
+fn infer_file_schema(
+    mut file: File,
+    delimiter: u8,
+    max_read_records: Option<usize>,
+    has_headers: bool,
+) -> Result<Schema> {
+    let mut csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone()?));
+
+    // get or create header names
+    // when has_headers is false, creates default column names with column_ prefix
+    let headers: Vec<String> = match has_headers {
+        true => {
+            let headers = &csv_reader.headers()?.clone();
+            headers.iter().map(|s| s.to_string()).collect()
+        }
+        false => {
+            let first_record_count = &csv_reader.headers()?.len();
+            (0..*first_record_count)
+                .map(|i| format!("column_{}", i + 1))
+                .into_iter()
+                .collect()
+        }
+    };
+
+    // save the csv reader position after reading headers
+    let position = csv_reader.position().clone();
+
+    let header_length = headers.len();
+    // keep track of inferred field types
+    let mut column_types: Vec<HashSet<DataType>> = vec![HashSet::new(); header_length];
+    // keep track of columns with nulls
+    let mut nulls: Vec<bool> = vec![false; header_length];
+
+    // return csv reader position to after headers
+    csv_reader.seek(position)?;
+
+    let mut fields = vec![];
+
+    for result in csv_reader
+        .into_records()
+        .take(max_read_records.unwrap_or(std::usize::MAX))
+    {
+        let record = result?;
+
+        for i in 0..header_length {
+            let string: Option<&str> = record.get(i);
+            match string {
+                Some(s) => {
+                    if s == "" {
+                        nulls[i] = true;
+                    } else {
+                        column_types[i].insert(infer_field_schema(s));
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    // build schema from inference results
+    for i in 0..header_length {
+        let possibilities = &column_types[i];
+        let has_nulls = nulls[i];
+        let field_name = &headers[i];
+
+        // determine data type based on possible types
+        // if there are incompatible types, use DataType::Utf8
+        match possibilities.len() {
```

Comment:
can we add `bool` to the list of types that we can infer?

Reply:
It's already there. In the case where a `bool` is included with another type, we default to UTF-8.